### ISSUE TITLE
Use Ada or ExUnits for script execution costs?

### ISSUE THREAD

**Issue by polinavino** [input-output-hk/cardano-ledger-specs#1048]
Do we want to only indicate the amount of Ada that running a script costs (or all the scripts in a transaction, if we go that route), or do we want to indicate this value in exunits? 

- The cost has two parameters, the amounts of resources used, and the "prices" of those resources per unit, so
scriptFee :: ExUnits -> CostMod -> Ada  

My thinking is that requiring it to be stated in Ada would charge the script interpreter with the task of doing the `scriptFee` computation, updating it at *each step* of the execution in order to compare it to the remaining fees value, whereas using `exunits` will result in a single computation on the ledger rules level to determine the Ada cost.

**Comment by michaelpj** [comment]
> My thinking is that requiring it to be stated in Ada would charge the script interpreter with the task of doing the scriptFee computation, updating it at each step of the execution in order to compare it to the remaining fees value

This is a multiplication and a subtraction, so essentially free.

And again, when we say "the interpreter" we just mean "some function". The actual interpreter that e.g. the Plutus team provides is going to work with whatever its execution units are. We're already going to have to check after each step whether we're out of resources, it costs very little more to do a multiplication and a subtraction and then compare against zero.

> whereas using exunits will result in a single computation on the ledger rules level to determine the Ada cost.

This also *requires* the limits to be listed in `exunits` somewhere, since we can't go from `Ada` to `exunits` because we don't know how it will be divided up.

**Comment by JaredCorduan** [comment]
using Ada instead of exunits seems like a less disruptive change to the existing spec, at least in terms of new types and case distinctions, etc.  but it does come at the cost of some extra state to be tracker in the `LEDGERS` transition, namely "fees left to fund the scripts"', but this seems pretty trivial.

**Comment by polinavino** [comment]
I am now more convinced we should indeed have a per-transaction fee which includes all scripts and iterates over them as each one spends part of the fee. Mostly due to the issue of signing `exunits`, which would allow us to get rid of the need for that, as well as get rid of the problem of having `exunits` be indexed by different things for different purposes (utxo spending vs certificates vs reward withdrawals)

**Comment by reactormonk** [comment]
On the other hand, I like the idea of having a pair of `(exunits, Ada)` which specifies how much I will pay for the transaction and how much computation power I want to buy for the `Ada`.

**Comment by michaelpj** [comment]
> Mostly due to the issue of signing exunits, which would allow us to get rid of the need for that

I have something to say about this, but I think it fits more in https://github.com/input-output-hk/cardano-ledger-specs/issues/1047, so I'll reply there.

> On the other hand, I like the idea of having a pair of (exunits, Ada) which specifies how much I will pay for the transaction and how much computation power I want to buy for the Ada.

Given that the "cost of computational power" is fixed by the cost model (the protocol parameters), I'm not sure what having both adds: they're just related by a known multiplication.

**Comment by reactormonk** [comment]
Gotcha. My suggestion would include removing said protocol parameter, and leave the cost model factor up to the transmitter.

**Comment by michaelpj** [comment]
Polina has pointed out an issue with having script execution costs in Ada. 

Here is a situation we have to consider: suppose a I submit a transaction in Epoch 1 that is only validated in Epoch 2. Since the *actual* cost of script execution varies with the protocol parameters, the cost when the transaction is validated may not match what the submitter expected. This will be a Phase 2 validation failure, so they'll lose their fees. Unpredictable Phase 2 validation failures are Bad!

How can we fix this? Well, the thing the user *does* know is how many `exunits` their scripts use. So if the limits were in `exunits`, then we can check in Phase 1 that there is sufficient fee to cover the stated number of `exunits` at the current price, and if not, we fail. So we fail in the same circumstances, but in Phase 1, which is Okay.

This is pretty annoying, but I don't have a better idea. There's various stuff we could do to try and ameliorate the problem, e.g. advertising protocol parameter changes for longer, allowing old parameters to be usable for a few epochs, etc., but I think this change would nail the problem, at the cost of going back to `exunits`.

**Comment by michaelpj** [comment]
(Incidentally, there's a nice ledger design principle we can take from this: Phase 2 validation should not refer to the protocol parameters at all, otherwise we open ourselves to this kind of variability!)

**Comment by michaelpj** [comment]
Quick reminder of some of the things that are bad about using `exunits` explicitly;
- `exunits` is likely very multidimensional, so they will physically take up a lot more space in the transaction, especially if we count all different builtins separately.
- It exposes implementation details of the evaluator, which nodes may then use to try and make decisions.

**Comment by polinavino** [comment]
Passing protocol parameters to the interpreter is definitely bad, that's a good overall rule. 

Note that there are 2 layers of indirection in calculating how much a transaction must pay for script execution. `ExUnits` or Mana should be an intermediate value that is calculated according to the cost model (this is NOT how it is in the spec I have been working on - this will be reconciled once we have a definitive decision here).
To clear up any confusion about `ExUnits`, here is a transcript of me asking @dcoutts about it:

Do we want two separate calculations:
* primitive operations -> exunits (this is what should be called the cost model)
Yes. The details of this is obviously very specific to the language.
* exunits -> Ada (this is called.. prices? i guess)
"Yes."

Do we want to be able to modify the coefficients in both calculations via protocol parameters?
"Yes.

If we do parametrize script validation by resource use, rather than Ada,
 should that be passed to the interpreter as primitives or `exunits`
"It'd be `exunits`."

The cost model (with its adjustable params) maps the execution (or execution trace) into execution cost units. The limit is in terms of execution cost units (or mapping of that to ADA if we decide to do that).

**Comment by polinavino** [comment]
So, i see two options here.. either we 
- figure out a way to restrict validation in a way that phase 1 would catch any problems related to cost model or price changes, and pass the cost model along with `exunits` or Ada to the interpreter, 
or
- smush the indirection into one, and have only one set of coefficients for the conversion 
`primitives -> Ada`
and pass the primitives or resource usage to the interpreter directly

**Comment by michaelpj** [comment]
> Note that there are 2 layers of indirection in calculating how much a transaction must pay for script execution. ExUnits or Mana should be an intermediate value that is calculated according to the cost model (this is NOT how it is in the spec I have been working on - this will be reconciled once we have a definitive decision here).

> smush the indirection into one, and have only one set of coefficients for the conversion
primitives -> Ada
and pass the primitives or resource usage to the interpreter directly
^^ I am guessing we can't smush the two because of the language specific differences in resource use, as Duncan said

Hm, I think we've been talking somewhat at cross-purposes about what the "cost model" is. I had thought that we didn't have this indirection, and that the cost model was directly "primitives -> Ada" as you put it. I hadn't thought about the idea that we might want our execution units to be common across different kinds of script languages.

Let's try and pick some consistent terminology: my understanding was that the "cost model" is the thing that we can control with the protocol parameters and which tells us how much things actually cost in Ada, i.e. the "exunits -> Ada" bit. I don't think we have a name for the way that the script interpreter calculates `exunits`, I'm going to call it the "execution tracker".

The problem with having some exposed `exunits` type that we can use in common across different script languages is that it potentially makes the cost model less useful. Suppose we used the `exunits` that Polina originally suggested, i.e. "execution steps" and "memory". Then we have *two* knobs for adjusting the cost model. We have no way of e.g. making "bytestring concatenation" specifically cost more, because we've already lost that information.

Can't we just change the execution tracker via the protocol parameters? No! That would lead to precisely the problem we discussed above, where the execution unit costs of a script would depend on the protocol parameters. So we can only change this along with a Plutus language version change. (Sounds like I disagree with Duncan about this!)

I also question whether we can have a meaningful `exunits` type that works across different script types. It seems likely that Plutus and multisig scripts won't do the same amount of work for one "execution step", and so any price for execution steps will over- or under-price one of them.

---

So much for being negative. Here's a positive suggestion. Let's have the `exunits` type be shared, but extensible. e.g.
```
type ExUnits = Map String Integer
```
The *cost model* can then *also* be extensible, i.e.
```
type CostModel = Map String Ada
```
So if we want to track "bytestring concatenation" for Plutus scripts, we can use `"Plutus-bytestring-concat"`as a key in those maps. (Of course, we have to do something if a unit doesn't have a cost in the cost model: I suggest a validation failure)

This also lets us somewhat kick the can down the road in terms of how multi-dimensional we want the `exunits` to be: we can add more kinds of thing over time.

Kind of ugly, but maybe it would work?

---

> For option 1, I think it could make sense to make the relevant parameters part of the transaction, so that they would be checked against the real current protocol parameters.

I also quite like this suggestion. It's a bit like the validity interval on transactions: you say "only validate this if the protocol parameters satisfy some property" and then you know during validation that they must satisfy that. We could also let people set inequality limits, so e.g they'd be happy if the costs go down, but not up.

**Comment by michaelpj** [comment]
Simon made the excellent point that if costs for builtin evaluations depend on the sizes of their arguments, which they might well, then the model I proposed above can't work, since it assumes we're just counting invocations.

**Comment by michaelpj** [comment]
So I'm inclining more towards the suggestion of constraining the parameters on the transaction.

**Comment by polinavino** [comment]
I will do it this way ^^ and see how it comes out. Should be an easy change.

**Comment by michaelpj** [comment]
A simple way of doing this would be to have an optional field that's the hash of the protocol parameters, and if it's present then it must match the current parameters. Inflexible, but if protocol parameter changes are infrequent, it might be fine.

**Comment by polinavino** [comment]
Should we make sure the real PP's get hashed only once during the ledger rules application for a single block (rather than for every transaction)?

**Comment by michaelpj** [comment]
For sure!

**Comment by polinavino** [comment]
This is settled - use ExUnits for execution budget, hash the cost model PPs
